fl_training.py

```diff
@@ -205,13 +205,266 @@ def build_aggregated_model(new_weights):
+    """Return a list with the aggregated model duplicated for each partner"""
+
+    aggregation_weights = prepare_aggregation_weights(aggregation_weighting_approach, partners_list, local_score_list)
+    partners_model_list = [None] * len(partners_list)
+    for partner_index, partner in enumerate(partners_list):
+        partners_model_list[partner_index] = build_aggregated_model(
+            aggregate_model_weights(model_list, aggregation_weights)
+        )
+    return partners_model_list
+
+
+def init_with_agg_model(partners_list, aggregation_weighting_approach, local_score_list, model_list):
+    """Return a new model aggregating models passed as argument"""
+
+    aggregation_weights = prepare_aggregation_weights(aggregation_weighting_approach, partners_list, local_score_list)
+    new_agg_model = build_aggregated_model(aggregate_model_weights(model_list, aggregation_weights))
+
+    return new_agg_model
+
+
+def build_from_previous_model(previous_model):
+    """Return a new model initialized with weights of a given model"""
+
+    new_model_from_previous_model = utils.generate_new_cnn_model()
+    new_model_from_previous_model.set_weights(previous_model.get_weights())
+    new_model_from_previous_model.compile(
+        loss=keras.losses.categorical_crossentropy,
+        optimizer="adam",
+        metrics=["accuracy"],
+    )
+
+    return new_model_from_previous_model
+
+
+def collaborative_round_fit(model_to_fit, train_data, val_data, batch_size):
+    """Fit the model with arguments passed as parameters and returns the history object"""
+
+    x_train, y_train = train_data
+    history = model_to_fit.fit(
+        x_train,
+        y_train,
+        batch_size=batch_size,
+        epochs=1,
+        verbose=0,
+        validation_data=val_data,
+    )
+
+    return history
+
+
+def print_collaborative_round_partner_result(
+        partner, partner_index, partners_count, collaborative_round_indexes, validation_score):
+    """Print the validation accuracy of the collaborative round"""
+
+    epoch_index, epoch_count, minibatch_index, minibatch_count = collaborative_round_indexes
+
+    epoch_nb_str = f"Epoch {str(epoch_index).zfill(2)}/{str(epoch_count - 1).zfill(2)}"
+    mb_nb_str = f"Minibatch {str(minibatch_index).zfill(2)}/{str(minibatch_count - 1).zfill(2)}"
+    partner_id_str = f"Partner id #{partner.id} ({partner_index}/{partners_count})"
+    val_acc_str = f"{round(validation_score, 2)}"
+
+    logger.info(f"{epoch_nb_str} > {mb_nb_str} > {partner_id_str} > val_acc: {val_acc_str}")
+
+
+def update_iterative_results(partner_index, collaborative_round_indexes, fit_history, iterative_results):
+    """Update the results arrays with results from the collaboration round"""
+
+    epoch_index, epoch_count, minibatch_index, minibatch_count = collaborative_round_indexes
+    local_score_list, model_list, score_matrix, score_matrix_extended = iterative_results
```

Comment:
If it makes things simpler, all of these variable could be members of the fl_scenario class. So we won't have to pass them around functions like we have to do now.
What do you think ?

Reply:
Oh yes, much better. But we don't have a class. Should I create one?

Reply:
Yes ! We can transform fl_scenario to be a class.